### PR TITLE
Refactor round context

### LINKED ISSUE #131: Refactor Contexts
The [Context](https://reactjs.org/docs/context.html)'s used in the frontend are quite cluttered and would benefit from some refactoring. I propose the following changes to remedy this and provide better maintainability and testability throughout the frontend.

### Proposed Changes

- [X]  **Rename files**
  - Example: `src/providers/ContextProviders/PlayersContextProvider.tsx` → `src/contexts/players.ts`.
  - See #127.
- [X] **Cleanup**
  - An initial refactor to cleanup interface exposed by each context and their internals
    - Each context would export a custom hook (i.e. `usePlayers()`) that would internally call [`useContext()`](https://reactjs.org/docs/hooks-reference.html#usecontext):
      - This means that the Context object (returned by [`React.createContext()`](https://reactjs.org/docs/context.html#reactcreatecontext) no longer needs to be exported - helping to hide context internals
      - The value of the Context object will be initialised to be `undefined` - this goes hand-in-hand with the custom hook by ensuring that it is used within the scope of the respective [`Provider`](https://reactjs.org/docs/context.html#contextprovider). The hook will throw an error when used outside a provider. This improves developer experience by making behaviour of the contexts more predictable.
    - The internals of the hooks updated to use memoization with [`useCallback()`](https://reactjs.org/docs/hooks-reference.html#usecallback) and [`useMemo()`](https://reactjs.org/docs/hooks-reference.html#usememo) to reduce unnecessary re-renders.
  - See #132.
- [ ] **Convert internal state to [`useReducer()`](https://reactjs.org/docs/hooks-reference.html#usereducer)**
  - This change would offer a number of improvements for each context:
    - It should largely remove the various functions returned by each context replaced by a single `dispatch()`.
    - It would also make testing state transitions easier through the `reducer()` function defined for each context.
  - Breakdown:
    - [x] Refactor `players` context. See #142.
    - [x] Refactor `punchlines` context. See #140.
    - [ ] Refactor `round` context. See #144.
    - [x] Remove `useCrud` hook. See #143.

### Future

While these changes will be a great improvement when complete, long-term I think it might be worth switching to [Redux](https://redux-toolkit.js.org) to replace these contexts.

Converting the contexts to [`useReducer()`](https://reactjs.org/docs/hooks-reference.html#usereducer) internally will be a great stepping stone for an eventual switch to Redux, as the rest of the application will already be configured to use `dispatch()` for state changes.

### PR DESCRIPTION
**Pull request by taitfuller**
This a child issue of #131. See the parent issue for greater context (pun intended) on the rationale for this change.

### Summary of changes

- Converts `RoundProvider` to `useReducer()` internally.
  - The reducer includes five actions: `NEW_ROUND`, `SET_SETUP`, `INCREMENT_PLAYERS_CHOSEN`, `SET_CHOSEN_PUNCHLINES`, `MARK_PUNCHLINE_READ`, and `SET_WINNER`.
  - This removes the following callbacks:
    - `setRoundNumber()` is replaced by the `NEW_ROUND` action.
    - `setSetup()` is replaced by the `SET_SETUP` action.
    - `incrementPlayersChosen()` is replaced by the `INCREMENT_PLAYERS_CHOSEN` action.
    - `setPunchlinesChosen()` is replaced by the `SET_CHOSEN_PUNCHLINES` action.
    - `markPunchlineRead()` is replaced by the `MARK_PUNCHLINE_READ` action.
    - `setHostViewIndex()` was not used within the code and was only removed.
    - `setWinner()`was replaced by the `SET_WINNER` action.
  - `RoundProvider` now returns a tuple consisting of the state object and dispatch function.
    - The state object contains keys for `roundNumber`, `setup`, `numPlayersChosen`, `chosenPunchlines`, and `winner`.
      - `roundNumber` is the number for the current round and is set by `NEW_ROUND`.
      - `setup` is the setup card (i.e. "Why did the chicken cross the road?") for the current round and is set by `SET_SETUP`.
      - `numPlayersChosen` is the number of players that have chosen a punchline for the current round. Each time a player choses a punchline it is incremented by `INCREMENT_PLAYERS_CHOSEN`.
      - `chosenPunchlines` is an array of `Punchline`'s chosen by all players.
      - `winner` is the winner of the round consisting of `winningPlayerId` for the winning players ID and `winningPunchlines` consisting of an array of strings for their winning punchlines (an array to support #81 in the future).
  - Improves testing to include 100% coverage for `src/contexts/round.tsx`

### Notes

- Some of the actions used in the reducer (i.e. `NEW_ROUND` and `SET_SETUP`) could potentially be combined where they always occur together. This has not been implemented in this PR as it would require backend changes which is considered outside of scope.
- Closes #131.

### CODE REVIEW

File: frontend/src/pages/EndRoundPage/index.tsx [hunk 1]
```diff
@@ -28,11 +28,11 @@ const EndRoundPage = ({ roundLimit }: { roundLimit: number }) => {
       <div className={styles.endOfRound}>
         <div style={{ margin: `24px 0` }}>
           <h5>The Setup:</h5>
-          <h2>{setup.setup}</h2>
+          <h2>{setup?.setup}</h2>
```
Comment (Alxjlv) [inline frontend/src/pages/EndRoundPage/index.tsx:31]: What happens if this is actually null? I notice you handle these cases where you make nullability changes in other parts of the code but not here

File: frontend/src/types/index.ts [hunk 2]
```diff
@@ -9,3 +9,13 @@ export type Punchline = {
   new?: boolean;
   viewed?: boolean;
 };
+
+export type Setup = {
+  setup: string;
+  type: "PICK_ONE" | "PICK_TWO" | "DRAW_TWO_PICK_THREE";
```
Comment (Alxjlv) [inline frontend/src/types/index.ts:15]: RIP still need to get these other modes working

**Review by Alxjlv — CHANGES_REQUESTED** [review]
Just want to see if the new nullability on the end round page needs to be handled

File: frontend/src/types/index.ts (hunk 2 above)
Comment (taitfuller) [reply frontend/src/types/index.ts:15]: Yep 🥲

File: frontend/src/pages/EndRoundPage/index.tsx (hunk 1 above)
Comment (taitfuller) [reply frontend/src/pages/EndRoundPage/index.tsx:31]: So with the [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) it returns `undefined` which is just rendered as blank by React - realistically we should probably look into some more advanced handling of these edge-cases down the line but for now I'm happy that this is good enough

Comment (Alxjlv) [reply frontend/src/pages/EndRoundPage/index.tsx:31]: Alright - I guess that's relatively graceful. Down the line better checking would be good but not a must

**Review by Alxjlv — APPROVED** [review]

**Pull request merged**